src/core/net/dns_client.cpp

```diff
@@ -43,357 +43,778 @@
+    SuccessOrExit(error = Name::ReadName(*mQuery, offset, aNameBuffer, aNameBufferSize));
 
 exit:
-    return message;
+    return error;
 }
 
-Message *Client::CopyAndEnqueueMessage(const Message &aMessage, const QueryMetadata &aQueryMetadata)
+otError Client::ServiceResponse::GetServiceInfo(ServiceInfo &aServiceInfo) const
 {
-    otError  error       = OT_ERROR_NONE;
-    Message *messageCopy = aMessage.Clone();
+    // Search and read SRV, TXT records in Answer Section
+    // matching name from query.
+
+    return FindServiceInfo(kAnswerSection, Name(*mQuery, kNameOffsetInQuery), aServiceInfo);
+}
+
+otError Client::ServiceResponse::GetHostAddress(const char *  aHostName,
+                                                uint16_t      aIndex,
+                                                Ip6::Address &aAddress,
+                                                uint32_t &    aTtl) const
+{
+    return FindHostAddress(kAdditionalDataSection, Name(aHostName), aIndex, aAddress, aTtl);
+}
 
-    VerifyOrExit(messageCopy != nullptr, error = OT_ERROR_NO_BUFS);
+#endif // OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+
+//---------------------------------------------------------------------------------------------------------------------
+// Client
+
+const uint16_t Client::kAddressQueryRecordTypes[] = {ResourceRecord::kTypeAaaa};
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+const uint16_t Client::kBrowseQueryRecordTypes[]  = {ResourceRecord::kTypePtr};
+const uint16_t Client::kServiceQueryRecordTypes[] = {ResourceRecord::kTypeSrv, ResourceRecord::kTypeTxt};
+#endif
+
+const uint8_t Client::kQuestionCount[] = {
+    /* (0) kAddressQuery -> */ OT_ARRAY_LENGTH(kAddressQueryRecordTypes), // AAAA records
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+    /* (1) kBrowseQuery  -> */ OT_ARRAY_LENGTH(kBrowseQueryRecordTypes),  // PTR records
+    /* (2) kServiceQuery -> */ OT_ARRAY_LENGTH(kServiceQueryRecordTypes), // SRV and TXT records
+#endif
+};
+
+const uint16_t *Client::kQuestionRecordTypes[] = {
+    /* (0) kAddressQuery -> */ kAddressQueryRecordTypes,
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+    /* (1) kBrowseQuery  -> */ kBrowseQueryRecordTypes,
+    /* (2) kServiceQuery -> */ kServiceQueryRecordTypes,
+#endif
+};
 
-    SuccessOrExit(error = aQueryMetadata.AppendTo(*messageCopy));
-    mPendingQueries.Enqueue(*messageCopy);
+Client::Client(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mSocket(aInstance)
+    , mTimer(aInstance, Client::HandleTimer)
+{
+    static_assert(kAddressQuery == 0, "kAddressQuery value is not correct");
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+    static_assert(kBrowseQuery == 1, "kBrowseQuery value is not correct");
+    static_assert(kServiceQuery == 2, "kServiceQuery value is not correct");
+#endif
+}
 
-    mRetransmissionTimer.FireAtIfEarlier(aQueryMetadata.mTransmissionTime);
+otError Client::Start(void)
```

Comment:
Instead of having `aServerAddr` for each public query method, it should be better to specify the server address with the `Start()` method. And it will ease adding auto-selection of DNS-SD servers which are found in Network Data.

We can leave it to a separate PR which will add auto-selection of the DNS-SD server. Thoughts?

Reply:
Let's consider this for a future/separate PR. 

- On one hand, I like the idea of being able to specify the server address for each query (e.g., address resolution query may want to use a different server than DNS-SD?). 
- On other hand, I see the benefit of being able to set the server address once and then use it. 

Maybe we can have a model where we have an API to set a "default server sock address" and if the caller from query API does not pass an address, we fallback and use the default one.